File: public/js/game/model/gameUser.js
function GameUser(uid){
    this.uid = uid
    this.score = 0;
    this.health = 100;
    this.room = null;
    
    Bindable(this);
    
    this.addHealth = function(delta, silentEvent){
        this.health = Math.max(0, this.health + delta)
        if(delta < 0)
            Engine.sound.play('growl');
        if(!silentEvent)
            this.dispatch('health', this)
    }
    this.addScore = function(delta, silentEvent){
        this.score += delta
        if(delta > 0)
            Engine.sound.play('collect');
        if(!silentEvent)
            this.dispatch('score', this)
    }
    this.isOwner = function(){
        return this.room && this.room.owner().uid == this.uid;
    }
    this.itsMe = function(){
        return Model.user.uid == this.uid
    }
    this.toData = function(){
        return {uid: this.uid, 
                health: this.health, 
                score: this.score}
    }
    this.fromData = function(data){
        this.score = data.score
        this.health = data.health
    }
}

GameUser.instancesByUid = {}
GameUser.instantiate = function(data){
    var uid = data.uid
    var user = GameUser.instancesByUid[uid]
    if(!user)
        GameUser.instancesByUid[uid] = user = new GameUser(uid)
    user.fromData(data)
    return user
}

File: src/main/java/com/hellsepontus/model/GameUser.java
package com.hellsepontus.model;

import com.hellsepontus.commands.ICommandClient;

import java.util.HashMap;
import java.util.Map;

public class GameUser implements IJson {
    public int uid;
    public int score = 0;
    public int health = 100;
    
    private static int uidCounter = 0;
    private ICommandClient client;
    private Room room;

    private GameUser(){
    }
    
    public ICommandClient getClient(){ return client; }
    public void setClient(ICommandClient client){ this.client = client; }

    public Room getRoom() { return room; }
    public void setRoom(Room room) { this.room = room;}

    public static GameUser instantiate(){
        return instantiate(++uidCounter);
    }

    protected static GameUser instantiate(int uid) {
        GameUser user = new GameUser();
        user.uid = uid;
        return user;
    }

    public Map<String, Object> toData() {
        Map<String, Object> result = new HashMap<String, Object>();
        result.put("uid", this.uid);
        result.put("score", this.score);
        result.put("health", this.health);
        return result;
    }

    public void fromData(Map<String, Object> data) {
        throw new NoSuchMethodError();
    }
    
    public boolean isOwner(){
        return room != null && room.owner().uid == this.uid;
    }
}
